// @wordpress/dataviews/build/dataforms-layouts/regular/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FormRegular;
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _normalizeFields = require("../../normalize-fields");
var _getVisibleFields = require("../get-visible-fields");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function FormRegular({
  data,
  fields,
  form,
  onChange
}) {
  const visibleFields = (0, _element.useMemo)(() => (0, _normalizeFields.normalizeFields)((0, _getVisibleFields.getVisibleFields)(fields, form.fields, form.combinedFields)), [fields, form.fields, form.combinedFields]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalVStack, {
    spacing: 4,
    children: visibleFields.map(field => {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(field.Edit, {
        data: data,
        field: field,
        onChange: onChange
      }, field.id);
    })
  });
}
