// @wordpress/block-library/build/home-link/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = HomeEdit;
var _clsx = _interopRequireDefault(require("clsx"));
var _blockEditor = require("@wordpress/block-editor");
var _i18n = require("@wordpress/i18n");
var _data = require("@wordpress/data");
var _coreData = require("@wordpress/core-data");
var _element = require("@wordpress/element");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

const preventDefault = event => event.preventDefault();
function HomeEdit({
  attributes,
  setAttributes,
  context
}) {
  const homeUrl = (0, _data.useSelect)(select => {
    // Site index.
    return select(_coreData.store).getEntityRecord('root', '__unstableBase')?.home;
  }, []);
  const {
    __unstableMarkNextChangeAsNotPersistent
  } = (0, _data.useDispatch)(_blockEditor.store);
  const {
    textColor,
    backgroundColor,
    style
  } = context;
  const blockProps = (0, _blockEditor.useBlockProps)({
    className: (0, _clsx.default)('wp-block-navigation-item', {
      'has-text-color': !!textColor || !!style?.color?.text,
      [`has-${textColor}-color`]: !!textColor,
      'has-background': !!backgroundColor || !!style?.color?.background,
      [`has-${backgroundColor}-background-color`]: !!backgroundColor
    }),
    style: {
      color: style?.color?.text,
      backgroundColor: style?.color?.background
    }
  });
  const {
    label
  } = attributes;
  (0, _element.useEffect)(() => {
    if (label === undefined) {
      __unstableMarkNextChangeAsNotPersistent();
      setAttributes({
        label: (0, _i18n.__)('Home')
      });
    }
  }, [label]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_jsxRuntime.Fragment, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      ...blockProps,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)("a", {
        className: "wp-block-home-link__content wp-block-navigation-item__content",
        href: homeUrl,
        onClick: preventDefault,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.RichText, {
          identifier: "label",
          className: "wp-block-home-link__label",
          value: label,
          onChange: labelValue => {
            setAttributes({
              label: labelValue
            });
          },
          "aria-label": (0, _i18n.__)('Home link text'),
          placeholder: (0, _i18n.__)('Add home link'),
          withoutInteractiveFormatting: true,
          allowedFormats: ['core/bold', 'core/italic', 'core/image', 'core/strikethrough']
        })
      })
    })
  });
}
