// @wordpress/dataviews/build/components/dataform-combined-edit/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function Header({
  title
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalVStack, {
    className: "dataforms-layouts__dropdown-header",
    spacing: 4,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
      alignment: "center",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalHeading, {
        level: 2,
        size: 13,
        children: title
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalSpacer, {})]
    })
  });
}
function DataFormCombinedEdit({
  field,
  data,
  onChange,
  hideLabelFromVision
}) {
  var _field$children;
  const className = 'dataforms-combined-edit';
  const visibleChildren = ((_field$children = field.children) !== null && _field$children !== void 0 ? _field$children : []).map(fieldId => field.fields.find(({
    id
  }) => id === fieldId)).filter(childField => !!childField);
  const children = visibleChildren.map(child => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "dataforms-combined-edit__field",
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(child.Edit, {
        data: data,
        field: child,
        onChange: onChange
      })
    }, child.id);
  });
  const Stack = field.direction === 'horizontal' ? _components.__experimentalHStack : _components.__experimentalVStack;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [!hideLabelFromVision && /*#__PURE__*/(0, _jsxRuntime.jsx)(Header, {
      title: field.label
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(Stack, {
      spacing: 4,
      className: className,
      as: "fieldset",
      children: children
    })]
  });
}
var _default = exports.default = DataFormCombinedEdit;
